### ISSUE TITLE
Blank spots while scrolling

### ISSUE THREAD

**Issue by drcmda** [bvaughn/react-window#43]
I'm seeing a lot of white while scrolling (screen cap below) in a list of 3xx items. They're all connected items, though (as in they're bound to a store), does this make a difference? Are elements re-created or cloned in order to virtualize them? I don't have that problem in the react-window demos, and they seem to have tons of items more, i wonder what's causing it for me ... 🤔

![rw](https://user-images.githubusercontent.com/2223602/44152130-f7da79a2-a0a4-11e8-85f4-87ef9fe598f3.gif)

And a quick screenshot of the code that produces the list:

![screenshot 2018-08-15 at 16 11 46](https://user-images.githubusercontent.com/2223602/44152532-1482fe3e-a0a6-11e8-8e1c-9356b50020e4.png)

**Comment by bvaughn** [comment]
My first question about the above screen cap is: Is it running in development mode? Because DEV mode is _significantly_ slower than production (due to extra validations and warnings, unminified/unoptimized code, etc.)

How far ahead of the scroll handler you can get when quickly scrolling– (i.e. how much blank white space you see) is not really related to the number of items in the list. It's totally a matter of how fast the scroll handler can run when creating and inserting new DOM elements. I'm not sure how much overhead there is in the `@subscribe` function that "connects" your components, but I'd start by trying to move that out/up if possible so that each individual item is lighter weight. (You might also want to verify that your `ListItem`s are memoizing correctly and not re-rendering on each scroll because of something to do with the `@subscribe` function).

You're also doing some array manipulation and mapping. This should be pretty fast, but you might consider experimenting with a [scroll indicator](https://react-window.now.sh/#/examples/list/scrolling-indicators) to render a lighter-weight version of yours rows _while scrolling_ to save more time.

But in general, this is just a limitation of windowing. The only way to ensure that there is _no_ empty space when scrolling is to force scrolling to run on the main/UI thread. The downside of this is that longer renders will make the UI jerky and less responsive to user input. I think some blank space around the edges is a better trade off.

**Comment by drcmda** [comment]
@bvaughn Thanks, very helpful!

Just one more question that i have, i logged componentDidMount and indeed, items do re-mount all the time. Now they'd go through a redux-like subscribe that binds them to a context store. I could imagine this is some overhead. The problem i have is that the list can get huge, say 100.000 items. And they all change status (done: false -> true) asynchroneously. If i would bind the list control to the tasks, it would re-render countless of times, this is what i wanted to avoid when binding liste-items individually. Is there anything react-window could help me with? Like telling me which items are being displayed, etc? How do people in general solve such issues?

**Comment by bvaughn** [comment]
List does provide an event, `onItemsRendered`, which is called whenever the items rendered by the list change:
```js
function onItemsRendered({
  overscanStartIndex,
  overscanStopIndex,
  visibleStartIndex,
  visibleStopIndex
}) {
  // All index params are numbers.
}
```

You could use this event event to keep track of which items are currently visible (and maybe store that info in `state`).

You could also pass the subscribed data from your `ItemList` component to your `ListItem` via the `itemData` key. (You might find [this react-window integration](https://github.com/facebook/react-devtools/blob/master/plugins/Profiler/views/SnapshotRanked.js) useful to look at, although it's advanced.)

This might take a bit of playing around with to get right, but it should be possible 👍

**Comment by drcmda** [comment]
Awesome, thanks a lot, this will do! 😃

**Comment by bvaughn** [comment]
Keep me posted if you learn anything new that could help improve the docs or be added to a new Code Sandbox example etc 😄

**Comment by lalitmee** [comment]
@bvaughn, Quick question. So if I am facing this issue it means I should refactor my **row component** and I should not do any calculation or data manipulation in my **row component** for fast rendering or we can say for getting rid of white spots on scrolling a little bit fast in the list. :thinking: 

Please let me know. I think I am facing this issue because I am doing a little bit of calculation on the basis of the `row index`. Maybe. :innocent:

**Comment by librz** [comment]
Had the same issue, in my case, I decide to set overscanCount to 30 so that the white space won't appear if scroll is not super fast, at the same time, overscanCount isn't super large as to hurt performance too much.

Using scroll indicator everytime user scrolls just isn't ideal for user exp.

I do hope there's a parameter that indicates whether the current row is rendering or not. In that way, we can show rendered rows as is, only display scroll indicator if the item is still being rendered.

In other words, I wish there is a `isRendering` to replace `isScrolling`
